TT-Transformers version 2 modules -- MLP

gwangTT opened the pull request:
### Ticket
Close #33782 #34289 #34290

### Problem description
TT-Transformers have been helpful in support more than a dozen models so far. However, to scale to 100+ models, we need to address several key issues:
- if-else in forward() function not only make the code hard to read but also slows down execution
- `ModelArgs` in `models/tt_transformers/tt/model_config.py` is too big (several thousand lines of code) and complex
- not enough unit testing

TT-Transformers version 2 (TTTv2) is created to address the above key issues:
- decouple config and compute -- minimal if-else in the forward functions
- ModelArgs of each module is refactored to co-locate with module code -- increase readability and maintainability
- Add unit tests to cover all use cases by models 
- Add code coverage analysis to CI

### What's changed
This PR is the first module contribution to TTTv2. Please see `user_guide.md` in this PR for an introduction of TTTv2 in general as well as how to use the MLP module. 

Another thing to highlight is the left-shifted unit testing of this module to provide two kinds of coverage:
- code coverage: currently the unit tests provides >90% code coverage of the MLP modules
- model coverage: `test_mlp_1d.py` has collected test parameters (tensor shapes, dtypes, model configs) for all the TTTv1 models currently in tt-metal CI on WH devices (N150, N300, T3000, and Galaxy); thus this PR provides 100% coverage of the TTTv1 models (MLP module) in the nightly and fast coverage in the unit test pipelines
As we adding more modules to TTTv2, we will keep adding such left-shifted unit testing.

In the process of making TTTv2 MLP module, I filed the following issues to report bugs and feature requests:
- https://github.com/tenstorrent/tt-metal/issues/35362
- https://github.com/tenstorrent/tt-metal/issues/35363

### Checklist

- [x] `T3K unit tests`: 
  - https://github.com/tenstorrent/tt-metal/actions/runs/20534818315
  - https://github.com/tenstorrent/tt-metal/actions/runs/20535453227
- [x] `T3K nightly tests`: 
  - https://github.com/tenstorrent/tt-metal/actions/runs/20534819337
  - https://github.com/tenstorrent/tt-metal/actions/runs/20535457018
- [x] `Galaxy unit tests`: 
  - https://github.com/tenstorrent/tt-metal/actions/runs/20635762861
- [x] revert [f79e20c](https://github.com/tenstorrent/tt-metal/pull/35095/commits/f79e20c4ccc2e3cea2560ac748b1cfaab5999362) through 98ca0e065a984d17bd5a0f9031d3446e989cdad8
- [x] revert [bf76ca3](https://github.com/tenstorrent/tt-metal/pull/35095/commits/bf76ca3890162815ef45eaf8d1ff72df849802d5) through 077cfe9499e8674f40526485fe519beb8467cf4b
- [x] revert e97c236707c51b21d628aa3f62378effd77d6f8a through 4ddb87097f3b3f70f5b6cf222c1f8459ee2e1310
- [x] revert f19d23f3fcf36ac5b61d938bc1900dff82c41c8a through 86b3a060f16ab74117458f684da24a13377f0c64
- [x] revert [9cf17dd](https://github.com/tenstorrent/tt-metal/pull/35095/commits/9cf17ddff4ff21c5f2847a3650ca9bbd10987585) through a413d51fb31015a222db6d9f9b38f4ab23c725fb

gwangTT committed:
First cut at TTTv2 MLP modules and unit tests.

gwangTT committed:
First passing all test cases in `test_mlp_2d.py`.

gwangTT committed:
clean up test_mlp_1d.py and use random weight for unit tests.

gwangTT committed:
enhance ttnn_mesh_device by creating submeshes when not using whole set of devices on a system but still set_fabric_config on the whiole set.

gwangTT committed:
remove incorrect work-around code in auto_compose.py; add more meshes to test cases

gwangTT committed:
Reset fabric after ttnn_mesh_device is closed.

gwangTT committed:
Add one more test case to test_auto_compose.py

gwangTT committed:
improve test_mlp_1d.py by only initializing weights of the MLP reference module.

gwangTT committed:
Found bug in ttnn.linear(); added negative test case and work-around.

gwangTT committed:
code cleanup

gwangTT committed:
Add TG use patterns to unit test; due to DP in current CI; all TG use patterns are MLP1D; add support for code coverage

gwangTT committed:
Optimize mlp_1d API for 90% of the use case; improved LazyWeight, device sharing, and torch tensor sharing to speed up unit tests of MLP1D

gwangTT committed:
Add tighter type hints for MLP1DConfig

gwangTT committed:
Remove derive-able property from MLP1DConfig

gwangTT committed:
code cleanup; add a todo to LazyWeight -- will tackle LazyWeight in the next PR.

gwangTT committed:
Cleanup mlp_2d.py

gwangTT committed:
Refactor MLP2D to match the Zen of TTTv2; udpated the test cases as well.

gwangTT committed:
Add more test cases which now covers all the TTTv1 models that uses MLP module in the tt-metal CI

gwangTT committed:
Fix the bugs found by the additional test cases; enhanced LazyWeight to automatically pad the tensors to satisfy alignment constraint by ttnn.from_torch

gwangTT committed:
Add users_guide.md

gwangTT committed:
Add tests to CI

gwangTT committed:
[to be reverted] turn on write permission for cache generation

gwangTT committed:
[to be reverted] turn on write permission to generate cache and download checkpoints.

gwangTT committed:
Add pytest-cov to requirements-dev.txt to support code coverage analysis

gwangTT committed:
code cleanup

gwangTT committed:
[to be reverted] turn on relevant tests only

gwangTT committed:
Add test cases for auto_compose.py on Galaxy!

gwangTT committed:
Remove test case unintended for check-in

gwangTT committed:
Fix CI issues

gwangTT committed:
[to be reverted] remove HF_HUB_OFFLINE temporarily

gwangTT committed:
[to be reverted] increase timeout values temporarily

gwangTT committed:
Fix pytest command line option

gwangTT committed:
Revert "[to be reverted] turn on write permission for cache generation"

This reverts commit f79e20c4ccc2e3cea2560ac748b1cfaab5999362.

gwangTT committed:
Revert "[to be reverted] remove HF_HUB_OFFLINE temporarily"

This reverts commit e97c236707c51b21d628aa3f62378effd77d6f8a.

gwangTT committed:
Revert "[to be reverted] turn on write permission to generate cache and download checkpoints."

This reverts commit bf76ca3890162815ef45eaf8d1ff72df849802d5.

gwangTT committed:
Revert "[to be reverted] increase timeout values temporarily"

This reverts commit f19d23f3fcf36ac5b61d938bc1900dff82c41c8a.

gwangTT committed:
Revert "[to be reverted] turn on relevant tests only"

This reverts commit 9cf17ddff4ff21c5f2847a3650ca9bbd10987585.

gwangTT committed:
Force HF_HUB_OFFLINE

gwangTT committed:
fix galaxy unit test

gwangTT committed:
reduce timeout value

gwangTT committed:
do not skip from_model_args test right now. Will retire later.

gwangTT committed:
Add unit tests to `test_auto_compose` to cover edge cases on Galaxy

maksim-tsishkouski-epam commented:
🚀 Awesome job 🚀 

Some notes from my side (please keep in mind that I can probably don't see the full picture)

### What I don't like in current architecture:

---

### Using same custom config

Block config is usually the same or almost the same for different decoder blocks.
Seems like it's not convenient to use same config with different weight.
Now it looks to be something like this

```
block0_config = MLP1DConfig(block0_w1, block0_w2, block0_w3, **same_config_params)
block1_config = MLP1DConfig(block1_w1, block1_w2, block1_w3, **same_config_params)
block0_mlp = MLP1D.from_config(block0_config)
block1_mlp = MLP1D.from_config(block1_config)
```

I would prefer to see something like

```
config = MLP1DConfig(**same_config_params)
block0_mlp = MLP1D.from_config(config)
block1_mlp = MLP1D.from_config(config)
block0_mlp.get_device_weight()  # will be done implicitly while calling MyLLM.get_device_weight()
block1_mlp.get_device_weight()  # will be done implicitly while calling MyLLM.get_device_weight()
```

Suggestions would be covered in the points below

---

### Model construction

If I understand correctly, now it's planned to be done with `weight + code` (LazyWeight, MLP/MLPConfig are based on torch.Tensor).
I propose to do it with `config (shape) + code` and then load weights (as it's done in e.g. torch).

---

### LazyWeight

In my opinion, required parameters: only shape (used in other e.g. MLP1DConfig) - similar to [torch meta device](https://docs.pytorch.org/docs/stable/meta.html).
Sources for shape value:
- "from_torch" - good
- "set up explicitly" - makes sense only with "from_cache" or need to have torch.Tensor while `get_device_weight` (preferred option)
- "from_cache" - without explicit shape the weight won't be lazy (or we'll have double loading) - I couldn't find the functionality to infer shape from file without loading it fully

I would suggest to **specify shape explicitly** and have separate `get_device_weight` with optional parameter of torch.Tensor.
As an alternative, we can leave torch.Tensor field but make it optional (we want to load weights without knowing about torch).

---

### Weights loading (from higher to lower level modules).

I think it can be done recursively for each submodule of the current module with potential `weight_mapping` (we want to support models without renaming to some unified format).
Pseudo-code

```
def get_device_weight(self, state_dict: Optional[Dict[str, torch.Tensor]]):
    for subname, submodule in self.named_children():
        if state_dict is None:
            submodule.get_device_weight()
        else:
            mapped_name = self.weight_mapping[subname]
            sub_state_dict = {k[len(mapped_name)]: v for k, v in state_dict.items() if k.startswith(mapped_name)}
            submodule.get_device_weight(sub_state_dict)
```

---

### Things to keep in mind

Serialization and deserialization:
- do we want to support it (loading config from file)? (seems like yes - I saw some code for serialization in utils)
- what (and how) should be serialized
- where to store configs: in code, in json/yaml, within the repo or in separate storage (e.g. git-lfs, in that case should we support versioning?)?

Not sure if it would be useful but we can support methods with the same name but different signature using [functools.singledispatchmethod](https://docs.python.org/3/library/functools.html#functools.singledispatchmethod)

yieldthought reviewed (approved):
I really like @maksim-tsishkouski-epam's review - a great level of discussion to see on PRs.

I'm sympathetic to the "let's do it like torch" angle but in this specific case:
1. I do prefer weight loading and caching to be explicit as it is here with LazyWeight for now - we're doing non-trivial conversions of the torch tensors in some cases (fusing QKV, swizzling weights) that would be extremely awkward to hide behind a generic .just_load_the_weights_bro() call.
2. Perfect is the enemy of good here and I think it's time to stop refactoring and get this shipped. It will be a delight for humans and agents alike to use and that's the most important thing right now.

Ship it and keep on cooking!

gwangTT commented on models/common/modules/lazy_weight.py:
```diff
@@ -0,0 +1,346 @@
+
+    The source can be:
+    - A tensor directly (duck-typed, typically torch.Tensor)
+    - A callable that returns a tensor (for truly lazy loading)
```
obsolete comment

gwangTT replied:
[311f90d](https://github.com/tenstorrent/tt-metal/pull/35095/commits/311f90d3bba17827901e42e78cfe482ac70aa112)

gwangTT commented on models/common/tests/conftest.py:
```diff
@@ -1,65 +1,141 @@
+        num_devices = parent_shape[0] * parent_shape[1]
+        if fabric_config is None and num_devices >= 8:
+            fabric_config = ttnn.FabricConfig.FABRIC_1D_RING
+        else:
```
This else is too complex and incorrect in some of the failed conditions of the if-statement.

gwangTT replied:
[f614d06](https://github.com/tenstorrent/tt-metal/pull/35095/commits/f614d06af49d4c63372f0210a85e8ab679d4e042)

gwangTT commented:
Hey @maksim-tsishkouski-epam, thanks for the detailed feedback! Let's discuss in more details in a meeting. 

My high-level response is that you are raising a lot of good questions that I plan to address with future PRs (here is a brief roadmap but we can add to it based on our discussions). I have ideas (some more concrete than others) to address all the questions. I think we can agree that technically there are a number of solutions but the final shape of the solution should probably be decided over time by learning from the users and incorporating additional feedback -- I would not be surprised that we keep on making changes and updates as the AI domain continue to evolve.  

So, let's talk about the big ideas separately and focus on the module-level issues within this PR. We will continue to improve on this code beyond this PR in the next few weeks.

gwangTT committed:
Fix logic error in tt_ccl.py when cluster_axis == 0

gwangTT committed:
remove callable source in LazyWeight and correctly handle pad_value

gwangTT committed:
check dynamic value `seq_len` is divisible by prefill_len_cutoff

gwangTT committed:
Fix logic error with fabric_config

gwangTT commented:
>     * High: models/common/modules/tt_ccl.py:75 treats cluster_axis=0 as falsy, so axis‑0 collectives use the “no‑cluster‑axis” semaphore pool; TG paths (e.g., axis‑0 all‑reduce) can cross‑talk or deadlock. Use cluster_axis is None or an explicit sentinel.

Addressed in [bf1bb84](https://github.com/tenstorrent/tt-metal/pull/35095/commits/bf1bb84a3c677fd1b2848e1975b48bcdae02a30b)
 
>     * Medium: models/common/modules/lazy_weight.py:96 and models/common/modules/lazy_weight.py:135 enforce source.shape and never call callables, despite the documented callable‑source API. Passing a lambda will assert or feed a function into ttnn.from_torch.

Addressed in [311f90d](https://github.com/tenstorrent/tt-metal/pull/35095/commits/311f90d3bba17827901e42e78cfe482ac70aa112)

>     * Medium: models/common/modules/mlp/mlp_1d.py:272 and models/common/modules/mlp/mlp_2d.py:383 reshape with seq_len // prefill_len_cutoff without checking divisibility; non‑multiple seq_len will fail reshape. Consider padding or an explicit assert.

Addressed in [63c31d7](https://github.com/tenstorrent/tt-metal/pull/35095/commits/63c31d751f228d1d99e3fa2c06f1d9db1af9d94b)
 
>     * Low: models/common/tests/conftest.py:83 overwrites any supplied fabric_config with a default, so tests cannot request non‑default fabric settings even when passed via fixture params.

[f614d06](https://github.com/tenstorrent/tt-metal/pull/35095/commits/f614d06af49d4c63372f0210a85e8ab679d4e042)

>     * Low: models/common/modules/lazy_weight.py:84, models/common/modules/lazy_weight.py:213, and models/common/modules/lazy_weight.py:265 never apply pad_value

Addressed in [311f90d](https://github.com/tenstorrent/tt-metal/pull/35095/commits/311f90d3bba17827901e42e78cfe482ac70aa112)

gwangTT committed:
Remove unnecessary .contiguous() call

gwangTT committed:
Refactor test code

gwangTT committed:
Merge branch 'main' into tttv2_mlp_modules

gwangTT committed:
Exclude tttv2_modules from the common unit tests

gwangTT committed:
Fix typo from merging from main
